numba/np/ufunc/dufunc.py

```diff
@@ -231,6 +242,264 @@ def _compile_for_argtys(self, argtys, return_type=None):
         self._lower_me.libs.append(cres.library)
         return cres
 
+    def _install_ufunc_attributes(self, template) -> None:
+
+        def get_attr_fn(attr: str) -> Callable:
+
+            def impl(ufunc):
+                val = getattr(ufunc.key[0], attr)
+                return lambda ufunc: val
+            return impl
+
+        # ntypes/types needs "at" to be a BoundFunction rather than a Function
+        # But this fails as it cannot a weak reference to an ufunc due to NumPy
+        # not setting the "tp_weaklistoffset" field. See:
+        # https://github.com/numpy/numpy/blob/7fc72776b972bfbfdb909e4b15feb0308cf8adba/numpy/core/src/umath/ufunc_object.c#L6968-L6983  # noqa: E501
+
+        at = types.Function(template)
+        attributes = ('nin', 'nout', 'nargs', # 'ntypes', # 'types',
+                      'identity', 'signature')
+        for attr in attributes:
+            attr_fn = get_attr_fn(attr)
+            overload_attribute(at, attr)(attr_fn)
+
+    def _install_ufunc_methods(self, template) -> None:
+        self._install_ufunc_reduce(template)
+
+    def _install_ufunc_reduce(self, template) -> None:
+        at = types.Function(template)
+
+        @overload_method(at, 'reduce')
+        def ol_reduce(ufunc, array, axis=0, dtype=None, initial=None):
+            if not isinstance(array, types.Array):
+                msg = 'The first argument "array" must be array-like'
+                raise errors.NumbaTypeError(msg)
+
+            axis_int = isinstance(axis, types.Integer)
+            axis_int_tuple = isinstance(axis, types.UniTuple) and \
+                isinstance(axis.dtype, types.Integer)
+            axis_tuple_size = len(axis) if axis_int_tuple else 0
```

Comment:
Might be worth adding corner cases like `axis=None` (reduce over all axes) and `axis=()` (reduce over nothing, but this had a different type signature i.e. `types.Tuple` with zero size).

Reply:
Done